### DISCUSSION TITLE
Adding class/interface tagging

### DISCUSSION THREAD

**Discussion by wimulkeman** [laravel/framework#42499] [Ideas]
**Idea**
I would love to have a feature in Laravel that enables us as developers to auto tag all the classes that implement an interface or extend a certain class.

The behaviour of this would be comparable with [autoconfiguring tags](https://symfony.com/doc/current/service_container/tags.html#autoconfiguring-tags) from Symfony.

**Inner workings**
For a project a made a PoC (Proof of Concept) for this where I used the classmapper from Composer to scan all the classes within a namespace (the namespace could be provided through a configuration). If the class extends / implements the expected class/interface, its service would receive the provided tag.

In the code the user could do something like the following in their service provider / app:
```php
$this->app->tagInstanceof(BaseServiceTypeInterface::class, 'my-tag');

$this->app
    ->when(MyDynamicHandlerService::class)
    ->needs(BaseServiceTypeInterface::class)
    ->giveTagged('my-tag');

// ...

class MyDynamicHandlerService
{
    public function __construct(BaseServiceTypeInterface ...$handlers)
    {
        $this->handlers = $handlers;
    }

    public function handle(string $type, array $config)
    {
        foreach ($this->handlers as $handler) {
            if (! $handler->supports($type)) {
                continue;
            }

            return $handler->handle($config);
        }

        throw new \LogicException('Missing handler for type '.$type);
    }
}
```
The performance penalty of doing this on runtime seemed to be marginal (less then 0.1 sec on a large application (+1700 classes), but I would recommend to make in the PoC I introduced a caching mechanism which prevented rerunning the same code when nothing changed.

**Why would I use this**
The use case for me was that I was implementing a service which retrieved a couple of handlers following the strategy pattern. This functionality enabled me to quickly add and remove handlers without having to manually declare them in a service provider or config.

**Comment by mysiar** [comment]
I miss as well automatic tagging of services on the interface they implement

**Comment by wimulkeman** [comment]
Based on the number of upvotes it seems I am not the only one who would like to see this become availble. I will start working on a PR to offer this functionality to the Laravel framework.

**Reply by mmorenoregalado** [reply]
Really! Anything we can do to help?

**Comment by wimulkeman** [comment]
I have opened a pull request and awaiting approval. https://github.com/laravel/framework/pull/44164

**Comment by luninroman** [comment]
Anything on this? We need this badly

**Reply by wimulkeman** [reply]
The MR was refused in its current form. A new option would be to implement it as a function like `tag_interface`. I would rather not implement it as a package myself.

I can offer it as a MR for the function if that would be handy.

**Reply by wimulkeman** [reply]
I have opened a new MR where the tagging is done through a user land function https://github.com/laravel/framework/pull/45910

**Reply by luninroman** [reply]
thank you! I hope they'll listen to you this time:)

**Comment by pbourdet** [comment]
Been trying Laravel for fun and was disappointed to see that such a useful feature is not built-in 😮

**Comment by devantoine** [comment]
Any news on this?

Is there a package that does that? I could use reflection in the ServiceProvider but I'd rather not to.

**Reply by henzeb** [reply]
It's not automatically, as you have to specify everything, but this is what we have currently: https://laravel.com/docs/11.x/container#tagging

Not sure if that is what people are looking for though?
